File: plc4j/api/src/main/java/org/apache/plc4x/java/api/messages/PlcRequest.java
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/
package org.apache.plc4x.java.api.messages;

import org.apache.plc4x.java.api.messages.items.RequestItem;

import java.util.LinkedList;
import java.util.List;
import java.util.Optional;

/**
 * Base type for all messages sent from the plc4x system to a connected plc.
 */
public abstract class PlcRequest<REQUEST_ITEM extends RequestItem> implements PlcMessage {

    protected final List<REQUEST_ITEM> requestItems;

    public PlcRequest() {
        this.requestItems = new LinkedList<>();
    }

    public PlcRequest(List<REQUEST_ITEM> requestItems) {
        this.requestItems = requestItems;
    }

    public void addItem(REQUEST_ITEM readRequestItem) {
        getRequestItems().add(readRequestItem);
    }

    public List<REQUEST_ITEM> getRequestItems() {
        return requestItems;
    }

    public Optional<? extends REQUEST_ITEM> getRequestItem() {
        if (isMultiValue()) {
            throw new IllegalStateException("too many items " + getNumberOfItems());
        }
        if (isEmpty()) {
            return Optional.empty();
        }
        return Optional.of(getRequestItems().get(0));
    }

    public void setRequestItem(REQUEST_ITEM requestItem) {
        if (isMultiValue()) {
            throw new IllegalStateException("too many items " + getNumberOfItems());
        }
        addItem(requestItem);
    }

    public int getNumberOfItems() {
        return getRequestItems().size();
    }

    public boolean isMultiValue() {
        return getNumberOfItems() > 1;
    }

    public boolean isEmpty() {
        return getNumberOfItems() < 1;
    }

    public abstract static class Builder<REQUEST_ITEM> {

        Class firstType;

        boolean mixed = false;

        List<REQUEST_ITEM> requests = new LinkedList<>();

        void checkType(Class dataType) {
            if (firstType == null) {
                firstType = dataType;
            }
            if (firstType != dataType) {
                mixed = true;
            }
        }
    }
}

File: sandbox/plc4py/plc4py/api/messages/PlcRequest.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
from abc import abstractmethod
from dataclasses import dataclass, field
from typing import Union, List

from plc4py.api.messages.PlcField import PlcField
from plc4py.api.messages.PlcMessage import PlcMessage
from plc4py.utils.GenericTypes import GenericGenerator


class PlcRequest(PlcMessage):
    """
    Base type for all messages sent from the plc4x system to a connected plc.
    """


@dataclass
class PlcFieldRequest(PlcRequest):
    fields: List[PlcField] = field(default_factory=lambda: [])

    @property
    def field_names(self):
        return [field.name for field in self.fields]


@dataclass
class PlcReadRequest(PlcFieldRequest):
    """
    Base type for all messages sent from the plc4x system to a connected plc.
    """


class ReadRequestBuilder(GenericGenerator):
    @abstractmethod
    def build(self) -> PlcReadRequest:
        pass

    @abstractmethod
    def add_item(self, field_query: Union[str, PlcField]) -> None:
        pass
